### PR TITLE
Tidy up timing and scope of `QueryGrid`

### PR DESCRIPTION
**Pull request by labkey-tchad**
#### Rationale
QueryGrid has some crufty support for switching grid tabs. `TabbedGridPanel` is more appropriately scoped and should be used instead.
The tab switching is a bit unreliable in `TabbedGridPanel` though; it needs a better way to ensure that the grid has switched. Waiting for the old grid to go stale then waiting for the new one to be loaded seems to work well.

#### Related Pull Requests
* https://github.com/LabKey/biologics/pull/1163
* https://github.com/LabKey/sampleManagement/pull/844
* https://github.com/LabKey/inventory/pull/386

#### Changes
* Improve timing of switching tabs on `TabbedGridPanel`
* Remove redundant `GridTabBar` component
* Add new ways to append Locators and expose `ComponentFinder` locators

### CODE REVIEW

File: src/org/labkey/test/components/Component.java [hunk 1]
```diff
@@ -153,7 +153,7 @@ protected F getThis()
             return (F) this;
         }
 
-        protected final Locator buildLocator()
+        public final Locator buildLocator()
```
Comment (labkey-chrisj) [inline src/org/labkey/test/components/Component.java:156]: oh, this is nice.

**Review by labkey-chrisj — APPROVED** [review]
I'm curious to learn more about the whys and use cases for Caching vs. Immutable locators, but my curiosity on that doesn't hold these branches up

File: src/org/labkey/test/components/ui/OmniBox.java [hunk 2]
```diff
@@ -100,28 +85,23 @@ public OmniBox tabOut()
...
+        return new OmniBoxValueFinder(getDriver()).findAll(this);
     }
 
     public OmniBoxValue getValue(String expected)
     {
-        return new OmniBoxValue.OmniBoxValueFinder(getDriver()).withText(expected).find(this);
+        return new OmniBoxValueFinder(getDriver()).withText(expected).find(this);
     }
 
     public Optional<OmniBoxValue> getOptionalValue(String expected)
```
Comment (labkey-danield) [inline src/org/labkey/test/components/ui/OmniBox.java:96]: Is this used any more?

File: src/org/labkey/test/components/ui/OmniBox.java [hunk 3]
```diff
@@ -140,32 +119,34 @@ private boolean isOpen()
...
-        getWrapper().waitFor(()-> !isOpen(), 500);
+        if (isOpen())
+        {
+            elementCache().input.sendKeys(Keys.ESCAPE);
+            WebDriverWrapper.waitFor(() -> !isOpen(), "Didn't close omnibox popup", 500);
+        }
         return this;
     }
 
     public boolean isEditing()  // there is always an input; 'editing' means it currently has a value
```
Comment (labkey-danield) [inline src/org/labkey/test/components/ui/OmniBox.java:130]: Does this need to be public?

File: src/org/labkey/test/components/ui/OmniBox.java [hunk 4]
```diff
@@ -140,32 +119,34 @@ private boolean isOpen()
 
     private OmniBox close()
```
Comment (labkey-danield) [inline src/org/labkey/test/components/ui/OmniBox.java:120]: It doesn't look like this is used.

File: src/org/labkey/test/components/ui/OmniBox.java [hunk 5]
```diff
@@ -140,32 +119,34 @@ private boolean isOpen()
...
 
     public boolean isEditing()  // there is always an input; 'editing' means it currently has a value
     {
-        return !Locators.editingValue.findElement(this)
-                .getAttribute("value").isEmpty();
+        return !elementCache().input.getAttribute("value").isEmpty();
     }
 
-    public OmniBox stopEditing()    // click just outside the omnibox, above/center
+    public OmniBox stopEditing()
```
Comment (labkey-danield) [inline src/org/labkey/test/components/ui/OmniBox.java:135]: I know this is a legacy method, but it just feels a little strange to have as a public method. 
Wouldn't a better design be to have an appendFilter or addFilter method that would add to the existing filter and that method takes care of editing "life cycle" of the OmniBox?

File: src/org/labkey/test/components/ui/Pager.java [hunk 6]
```diff
@@ -160,7 +151,7 @@ public int total()
      * if the pager does not show a total, it indicates that there are no more to the set than are shown on the current page.
      * use this to understand whether or not there are more pages of information (which
      * could be the whole set, or a filtered range of it, or both).
-     * @return
+     * @return <code>true</code> if pager shows a "total"
      */
     public boolean hasTotal()
```
Comment (labkey-danield) [inline src/org/labkey/test/components/ui/Pager.java:156]: I'm going to make an argument to remove this method. It doesn't look like it is used (that should be reason enough), and second elementCache().counts has a time out of 4 seconds. If I understand the code correctly that means to get a false from this method you have to wait 4 seconds.

**Review by labkey-danield — APPROVED** [review]
You may already be aware of this but there are changes being planned for the OmniBox. I think around filtering and sorting. Not sure if they would impact the test component.

**Review by labkey-danield — COMMENTED** [review]
Please retarget these changes to 22.3.

**Pull request merged**